scripts/train/train.py

```diff
@@ -164,170 +165,315 @@ def build_dataloader(cfg: DictConfig, tokenizer: PreTrainedTokenizerBase,
     with init_context:
-        if cfg.get('lora',
-                   None) is not None:  # frozen model + trainable lora modules
+        if lora_config is not None:  # frozen model + trainable lora modules
             model: ComposerHFCausalLM = build_composer_peft_model(
-                cfg.model, cfg.lora, tokenizer)
+                model_config, lora_config, tokenizer)
             print_trainable_parameters(model)  # should not be 100%
         else:  # standard model
-            model = build_composer_model(cfg.model, tokenizer)
-    cfg.n_params = sum(p.numel() for p in model.parameters())
-    print(f'{cfg.n_params=:.2e}')
+            model = build_composer_model(model_config, tokenizer)
+
+    # Log number of parameters
+    n_params = sum(p.numel() for p in model.parameters())
+    log_config(om.create({'n_params': n_params}))
 
     # Dataloaders
     print('Building train loader...')
     train_loader = build_dataloader(
-        cfg.train_loader,
+        train_loader_config,
         tokenizer,
-        cfg.device_train_batch_size,
+        device_train_batch_size,
     )
+
+    ## Evaluation
     print('Building eval loader...')
     evaluators = []
-    if 'eval_loader' in cfg:
+    if eval_loader_config is not None:
         assert model.train_metrics is not None
+        eval_dataloader = build_dataloader(eval_loader_config, tokenizer,
+                                           device_eval_batch_size)
+        eval_metric_names = list(model.train_metrics.keys())
         eval_loader = Evaluator(label='eval',
-                                dataloader=build_dataloader(
-                                    cfg.eval_loader, tokenizer,
-                                    cfg.device_eval_batch_size),
-                                metric_names=list(model.train_metrics.keys()))
+                                dataloader=eval_dataloader,
+                                metric_names=eval_metric_names)
         evaluators.append(eval_loader)
 
-    if 'icl_tasks' in cfg:
-        icl_evaluators, _ = build_icl_evaluators(cfg.icl_tasks, tokenizer,
-                                                 cfg.max_seq_len,
-                                                 cfg.device_eval_batch_size)
+    if icl_tasks_config is not None:
+        icl_evaluators, _ = build_icl_evaluators(icl_tasks_config, tokenizer,
+                                                 max_seq_len,
+                                                 device_eval_batch_size)
         evaluators.extend(icl_evaluators)
 
     # Optimizer
-    optimizer = build_optimizer(cfg.optimizer, model)
+    optimizer = build_optimizer(optimizer_config, model)
 
     # Scheduler
-    scheduler = build_scheduler(cfg.scheduler)
+    scheduler = build_scheduler(scheduler_config)
 
     # Loggers
     loggers = [
-        build_logger(name, logger_cfg)
-        for name, logger_cfg in (cfg.get('loggers') or {}).items()
-    ]
+        build_logger(str(name), logger_cfg)
```

Comment:
why did this need a str cast?

Reply:
It's of type `DictKeyType` whereas build_logger takes type string